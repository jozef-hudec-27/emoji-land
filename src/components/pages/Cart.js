import { useState } from 'react'
import { useNavigate } from 'react-router-dom'
import CartEmpty from '../cart/CartEmpty'
import CartItems from '../cart/CartItems'
import CartSummary from '../cart/CartSummary'
import CheckoutModal from '../CheckoutModal'
import Page from './Page'

export default function Cart({ cart, cartContains, addToCart, changeQuantity, calculateTotalCart, clearCart }) {
  const [showCheckoutModal, setShowCheckoutModal] = useState(false)
  const navigate = useNavigate()

  return (
    <Page>
      <section id="cart" style={{ display: !cart.length && 'block' }}>
        {cart.length ? (
          <>
            <section id="cart-items-wrapper" className="flexbox flex-column flex-align-center gap-8">
              <h2>Cart items</h2>

              <CartItems
                items={cart}
                cartContains={cartContains}
                addToCart={addToCart}
                changeQuantity={changeQuantity}
              />
            </section>

            <section id="cart-summary-wrapper" className="flexbox flex-column flex-align-center gap-8">
              <h2>Summary</h2>

              <CartSummary
                items={cart}
                calculateTotalCart={calculateTotalCart}
                clearCart={clearCart}
                setShowCheckoutModal={setShowCheckoutModal}
              />
            </section>
          </>
        ) : (
          <CartEmpty />
        )}
      </section>

      <CheckoutModal
        show={showCheckoutModal}
        onXmark={() => setShowCheckoutModal(false)}
        onContinueShopping={() => navigate('/shop')}
      />
    </Page>
  )
}
