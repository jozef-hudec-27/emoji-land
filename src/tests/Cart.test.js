import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { act } from 'react-dom/test-utils'
import { BrowserRouter } from 'react-router-dom'
import Cart from '../components/pages/Cart'

describe('Cart component', () => {
  it('renders empty cart message when no items are in cart', () => {
    render(
      <BrowserRouter>
        <Cart cart={[]} />
      </BrowserRouter>
    )

    expect(screen.getByRole('heading', { name: 'Your cart is empty. Go and buy some emojis!' })).toBeInTheDocument()
  })

  it('matches snapshot if cart is not empty', () => {
    const cart = [
      [
        {
          emoji: 'ðŸ˜€',
          price: 1,
          name: 'Grinning Face',
          description:
            'A yellow face with simple, open eyes and a broad, open smile, showing upper teeth and tongue on some platforms. Often conveys general pleasure and good cheer or humor.',
        },
        2,
      ],
    ]

    const { container } = render(
      <BrowserRouter>
        <Cart
          cart={cart}
          cartContains={jest.fn()}
          addToCart={jest.fn()}
          changeQuantity={jest.fn()}
          calculateTotalCart={jest.fn()}
          clearCart={jest.fn()}
        />
      </BrowserRouter>
    )

    expect(container).toMatchSnapshot()
  })

  it('clears cart after clicking button', () => {
    const cart = [
      [
        {
          emoji: 'ðŸ˜€',
          price: 1,
          name: 'Grinning Face',
          description:
            'A yellow face with simple, open eyes and a broad, open smile, showing upper teeth and tongue on some platforms. Often conveys general pleasure and good cheer or humor.',
        },
        2,
      ],
    ]
    const clearCart = jest.fn()

    render(
      <BrowserRouter>
        <Cart
          cart={cart}
          cartContains={jest.fn()}
          addToCart={jest.fn()}
          changeQuantity={jest.fn()}
          calculateTotalCart={jest.fn()}
          clearCart={clearCart}
        />
      </BrowserRouter>
    )

    const clearCartBtn = screen.getByRole('button', { name: 'Clear Cart' })
    userEvent.click(clearCartBtn)

    expect(clearCart).toHaveBeenCalled()
  })

  it('displays checkout modal after clicking the button', async () => {
    const cart = [
      [
        {
          emoji: 'ðŸ˜€',
          price: 1,
          name: 'Grinning Face',
          description:
            'A yellow face with simple, open eyes and a broad, open smile, showing upper teeth and tongue on some platforms. Often conveys general pleasure and good cheer or humor.',
        },
        2,
      ],
    ]
    const clearCart = jest.fn()

    render(
      <BrowserRouter>
        <Cart
          cart={cart}
          cartContains={jest.fn()}
          addToCart={jest.fn()}
          changeQuantity={jest.fn()}
          calculateTotalCart={jest.fn()}
          clearCart={clearCart}
        />
      </BrowserRouter>
    )

    const checkoutBtn = screen.getByRole('button', { name: 'Checkout' })
    act(() => {
      userEvent.click(checkoutBtn)
    })

    expect(clearCart).toHaveBeenCalled()

    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'Continue shopping' })).toBeInTheDocument()
    })
  })
})
