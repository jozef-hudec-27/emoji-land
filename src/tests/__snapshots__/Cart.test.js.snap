// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component matches snapshot if cart is not empty 1`] = `
<div>
  <div
    class="flexbox flex-justify-center full-height p-8"
  >
    <section
      id="cart"
    >
      <section
        class="flexbox flex-column flex-align-center gap-8"
        id="cart-items-wrapper"
      >
        <h2>
          Cart items
        </h2>
        <div
          class="cart-items flexbox flex-center flex-wrap gap-12"
        >
          <div
            class="cart-item card"
          >
            <div
              class="flexbox flex-column flex-center"
            >
              <p
                style="font-size: 4rem;"
              >
                ðŸ˜€
              </p>
              <div
                class="product-counter flexbox flex-align-center ml-auto gap-10"
              >
                <button
                  aria-label="Increment"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-chevron-up "
                    data-icon="chevron-up"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 448 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M201.4 105.4c12.5-12.5 32.8-12.5 45.3 0l192 192c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L224 173.3 54.6 342.6c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3l192-192z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
                <p>
                  2
                </p>
                <button
                  aria-label="Decrement"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-chevron-down "
                    data-icon="chevron-down"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 448 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </section>
      <section
        class="flexbox flex-column flex-align-center gap-8"
        id="cart-summary-wrapper"
      >
        <h2>
          Summary
        </h2>
        <div
          class="cart-summary full-width"
        >
          <table
            class="cart-summary-table"
          >
            <thead>
              <tr>
                <th>
                  Item
                </th>
                <th>
                  Quantity
                </th>
                <th>
                  Price ($)
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  ðŸ˜€
                </td>
                <td>
                  2
                </td>
                <td>
                  1
                </td>
              </tr>
            </tbody>
          </table>
          <div
            class="flexbox flex-center gap-16 mt-12"
          >
            <h3>
              Total: $
            </h3>
            <div
              class="flexbox gap-4"
            >
              <button
                class="clear-cart-btn"
              >
                Clear Cart
              </button>
              <button
                class="checkout-btn"
              >
                Checkout
              </button>
            </div>
          </div>
        </div>
      </section>
    </section>
  </div>
</div>
`;
